### PR TITLE
Implement -R option for dvc metrics show

### PR DESCRIPTION
**Pull request by tdeboissiere**
Solves #1465

### CODE REVIEW

File: dvc/command/metrics.py [hunk 1]
```diff
@@ -32,8 +32,10 @@ def run(self):
                                       typ=typ,
                                       xpath=xpath,
                                       all_branches=self.args.all_branches,
-                                      all_tags=self.args.all_tags)
-        except DvcException:
+                                      all_tags=self.args.all_tags,
+                                      recursive=self.args.recursive)
+        except DvcException as e:
+            print(e)
```
Comment (shcheklein) [inline dvc/command/metrics.py]: remove print. logger.error should handle this.

File: dvc/command/metrics.py [hunk 2]
```diff
@@ -139,6 +141,12 @@ def add_parser(subparsers, parent_parser):
         action='store_true',
         default=False,
         help='Show metrics for all tags.')
+    metrics_show_parser.add_argument(
+        '-R',
+        '--recursive',
+        action='store_true',
+        default=False,
+        help='Show metrics recursively.')
```
Comment (shcheklein) [inline dvc/command/metrics.py]: we should probably update the position argument description - it accepts DVC file or a directory now, right?

Comment (shcheklein) [reply dvc/command/metrics.py]: Can we do something like: 'Recursively search and process metric files in path.'? Or something else, just to make it a little bit more clear what's happening here.

File: dvc/project.py [hunk 3]
```diff
@@ -1128,35 +1160,39 @@ def metrics_show(self,
...
+        outs = self._find_output_by_path(path)
+
+        if not outs:
             msg = "unable to find file '{}' in the pipeline".format(path)
             raise DvcException(msg)
 
-        if out.scheme != 'local':
-            msg = "output '{}' scheme '{}' is not supported for metrics"
-            raise DvcException(msg.format(out.path, out.path_info['scheme']))
+        for out in outs:
```
Comment (shcheklein) [inline dvc/project.py]: do we expect modify to deal with multiple metric files as well? I mean, do we actually need a cycle here?

File: tests/test_metrics.py [hunk 4]
```diff
@@ -8,253 +8,343 @@
 
 
 class TestMetrics(TestDvc):
+
     def setUp(self):
         super(TestMetrics, self).setUp()
-        self.dvc.scm.commit('init')
+        self.dvc.scm.commit("init")
```
Comment (shcheklein) [inline tests/test_metrics.py]: Why this change in tests? ' to "?

File: dvc/project.py [hunk 5]
```diff
@@ -1041,47 +1041,74 @@ def _read_metric(self, path, typ=None, xpath=None):
...
+            if len(stages) > 1:
+                raise OutputDuplicationError(path, stages)
+        else:
+            matched = [out for out in outs if out.path == abs_path]
+            stages = [out.stage.relpath for out in matched]
+            if len(stages) > 1:
+                raise OutputDuplicationError(path, stages)
 
-        return matched[0] if matched else None
+        return matched if matched else None
```
Comment (shcheklein) [inline dvc/project.py]: I would return an empty list instead of None here. It can spare some if's upstream.

File: dvc/project.py [hunk 6]
```diff
@@ -1041,47 +1041,74 @@ def _read_metric(self, path, typ=None, xpath=None):
...
             outs = [out for stage in astages for out in stage.outs]
 
             if path:
-                out = self._find_output_by_path(path, outs=outs)
-                stage = out.stage.path if out else None
-                if out and all([out.metric,
+                outs = self._find_output_by_path(path,
+                                                 outs=outs,
+                                                 recursive=recursive)
+                stages = [out.stage.path for out in outs] if outs else None
```
Comment (shcheklein) [inline dvc/project.py]: if is not required if you we can return [] instead of None here

File: dvc/project.py [hunk 7]
```diff
@@ -1041,47 +1041,74 @@ def _read_metric(self, path, typ=None, xpath=None):
...
+                entries = []
+                if outs:
```
Comment (shcheklein) [inline dvc/project.py]: same ^^

File: dvc/project.py [hunk 8]
```diff
@@ -1041,47 +1041,74 @@ def _read_metric(self, path, typ=None, xpath=None):
...
-        matched = [out for out in outs if out.path == abs_path]
-        stages = [out.stage.relpath for out in matched]
-        if len(stages) > 1:
-            raise OutputDuplicationError(path, stages)
+        if os.path.isdir(abs_path) and recursive:
+            matched = [out for out in outs
+                       if os.path.abspath(out.path).startswith(abs_path)]
+            stages = [out.stage.relpath for out in matched
+                      if out.path == abs_path]
+            if len(stages) > 1:
```
Comment (shcheklein) [inline dvc/project.py]: can it be > 1 in this case?

File: tests/test_metrics.py (hunk 4 above)
Comment (tdeboissiere) [reply tests/test_metrics.py]: Will revert to original. I generally use black for formatting, and it automatically converts ' to "

Comment (shcheklein) [reply tests/test_metrics.py]: @tdeboissiere confirm this @efiop, btw. It looks like we are migrating to black, so may be should keep it your way.

Comment (efiop) [reply tests/test_metrics.py]: @shcheklein We've discussed it with @tdeboissiere on discord :slightly_smiling_face:  Let's merge https://github.com/iterative/dvc/pull/1547 first and then rebase this PR on top of it, so we do everything properly :slightly_smiling_face:

Comment (shcheklein) [reply tests/test_metrics.py]: 👍

**Comment by efiop** [comment]
@tdeboissiere https://github.com/iterative/dvc/pull/1547 is merged. Please rebase :slightly_smiling_face:

File: dvc/project.py (hunk 8 above)
Comment (tdeboissiere) [reply dvc/project.py]: It should not be. If path is a folder, then it should not show up as an element of outs.
Removing

File: dvc/project.py (hunk 3 above)
Comment (tdeboissiere) [reply dvc/project.py]: Yes, because outs comes from `self._find_output_by_path(path)`, which in case the -R is active and multiple metric files have been found will return a list

Comment (shcheklein) [reply dvc/project.py]: @tdeboissiere totally! My question was more from a product perspective - what should we do when someone runs `dvc metric modify [something] -R path`. Should we allow that at all or just raise an exception? Should we care for this operation to be atomic? Probably, @efiop has more knowledge on this.

File: dvc/command/metrics.py [hunk 9]
```diff
@@ -133,6 +136,13 @@ def add_parser(subparsers, parent_parser):
         default=False,
         help="Show metrics for all tags.",
     )
+    metrics_show_parser.add_argument(
+        "-R",
+        "--recursive",
+        action="store_true",
+        default=False,
+        help="If path is a folder, recursively search and process metric files in path.",
```
Comment (shcheklein) [inline dvc/command/metrics.py]: ❤️ tx!

File: dvc/project.py (hunk 3 above)
Comment (efiop) [reply dvc/project.py]: @shcheklein is right, since -R is not supported for `metrics_modify` let's remove this cycle and put something like
```
assert(len(outs) == 1)
```
for now.

File: tests/test_metrics.py [hunk 10]
```diff
@@ -97,6 +97,63 @@ def test(self):
...
+    def setUp(self):
+        super(TestMetricsRecursive, self).setUp()
+        self.dvc.scm.commit("init")
+
+        self.dvc.scm.checkout("nested", create_new=True)
+
+        os.mkdir("nested")
+        os.mkdir("nested/subnested")
+
+        main(
```
[the remark this answers is not in the document]
Comment (tdeboissiere) [reply tests/test_metrics.py]: Done

File: tests/test_metrics.py [hunk 11]
```diff
@@ -97,6 +97,63 @@ def test(self):
...
 
+class TestMetricsRecursive(TestDvc):
+    def setUp(self):
+        super(TestMetricsRecursive, self).setUp()
+        self.dvc.scm.commit("init")
+
+        self.dvc.scm.checkout("nested", create_new=True)
+
+        os.mkdir("nested")
+        os.mkdir("nested/subnested")
```
[the remark this answers is not in the document]
Comment (tdeboissiere) [reply tests/test_metrics.py]: Done

File: tests/test_metrics.py [hunk 12]
```diff
@@ -97,6 +97,63 @@ def test(self):
...
+
+        main(
+            [
+                "run",
+                "-M",
+                "nested/metric_nested",
```
[the remark this answers is not in the document]
Comment (tdeboissiere) [reply tests/test_metrics.py]: Done

File: scripts/hooks/hook-google.cloud.py [hunk 13]
```diff
@@ -1,3 +1,4 @@
 from PyInstaller.utils.hooks import copy_metadata
-datas = copy_metadata('google-cloud-core')
-hiddenimports = ['google.api']
+
```
[the remark this answers is not in the document]
Comment (tdeboissiere) [reply scripts/hooks/hook-google.cloud.py]: Done

**Comment by tdeboissiere** [comment]
Hmm.. Not familiar with windows, I don't know why the test did not pass

File: tests/test_metrics.py [hunk 14]
```diff
@@ -97,6 +97,67 @@ def test(self):
...
+        self.dvc.scm.checkout("master")
+
+    def test(self):
+
+        ret = self.dvc.metrics_show("nested", all_branches=True, recursive=False)
+        self.assertEqual(len(ret), 0)
+
+        ret = self.dvc.metrics_show("nested", all_branches=True, recursive=True)
+        self.assertEqual(len(ret), 1)
+        self.assertTrue(
```
Comment (efiop) [inline tests/test_metrics.py]: Let's use `self.assertEqual()` here and down below instead of manually checking `==`. This will help us debug test failure on windows.

File: tests/test_metrics.py [hunk 15]
```diff
@@ -97,6 +97,67 @@ def test(self):
...
+
+        os.mkdir("nested")
+        os.mkdir(os.path.join("nested", "subnested"))
+
+        ret = main(
+            [
+                "run",
+                "-M",
+                os.path.join("nested", "metric_nested"),
+                "echo",
```
Comment (efiop) [inline tests/test_metrics.py]: Looks like `echo` on windows doesn't know about `-n` option :) Could we try without it? You might want to make `_read_metric` strip what it reads before showing https://github.com/iterative/dvc/blob/master/dvc/project.py#L1023 .

File: dvc/command/metrics.py [hunk 16]
```diff
@@ -98,7 +99,9 @@ def add_parser(subparsers, parent_parser):
         description=METRICS_SHOW_HELP,
         help=METRICS_SHOW_HELP,
     )
-    metrics_show_parser.add_argument("path", nargs="?", help="Path to metrics file.")
+    metrics_show_parser.add_argument(
+        "path", nargs="?", help="Path to metrics file or folder"
```
Comment (efiop) [inline dvc/command/metrics.py]: Let's use `directory` instead of `folder` here and down below since we use it in every other place.
